internal/configwatcher/watcher.go

```diff
@@ -0,0 +1,109 @@
+package configwatcher
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/facebookgo/startstop"
+	"github.com/honeycombio/refinery/config"
+	"github.com/honeycombio/refinery/internal/otelutil"
+	"github.com/honeycombio/refinery/pubsub"
+	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
+)
+
+const ConfigPubsubTopic = "cfg_update"
+
+// This exists in internal because it depends on both config and pubsub.
+// So we have to create it after creating pubsub and let dependency injection work.
+
+// ConfigWatcher listens for configuration changes and publishes notice of them.
+// It avoids sending duplicate messages by comparing the hash of the configs.
+type ConfigWatcher struct {
+	Config  config.Config
+	PubSub  pubsub.PubSub
+	Tracer  trace.Tracer `inject:"tracer"`
+	subscr  pubsub.Subscription
+	msgTime time.Time
+	done    chan struct{}
+	startstop.Starter
+	startstop.Stopper
+}
+
+// ReloadCallback is used to tell others that the config has changed.
+// This gets called whenever it has actually changed, but it might have
+// changed because we were told about it in pubsub, so we don't publish
+// a message if the hashes are the same.
+func (cw *ConfigWatcher) ReloadCallback(cfgHash, rulesHash string) {
+	ctx := context.Background()
+	ctx, span := otelutil.StartSpanMulti(ctx, cw.Tracer, "ConfigWatcher.ReloadCallback", map[string]any{
+		"new_config_hash": cfgHash,
+		"new_rules_hash":  rulesHash,
+	})
+	defer span.End()
+
+	// don't publish if we have recently received a message (this avoids storms)
+	now := time.Now()
+	if now.Sub(cw.msgTime) < time.Duration(cw.Config.GetGeneralConfig().ConfigReloadInterval) {
+		otelutil.AddSpanField(span, "sending", false)
+		return
+	}
+
+	message := now.Format(time.RFC3339)
+	otelutil.AddSpanFields(span, map[string]any{"sending": true, "message": message})
+	cw.PubSub.Publish(ctx, ConfigPubsubTopic, message)
+}
+
+// SubscriptionListener listens for messages on the config pubsub topic and reloads the config
+// if a new set of hashes is received.
+func (cw *ConfigWatcher) SubscriptionListener(ctx context.Context, msg string) {
+	_, span := otelutil.StartSpanWith(ctx, cw.Tracer, "ConfigWatcher.SubscriptionListener", "message", msg)
+	defer span.End()
+
+	// parse message as a time in RFC3339 format
+	msgTime, err := time.Parse(time.RFC3339, msg)
+	if err == nil {
```

Comment:
Is this a typo? should this be `err != nil`?

Reply:
Eeek! Thanks.